File: 30 Days of Code/Day 12 Inheritance/Day 12 Inheritance.cs
using System;
using System.Linq;

class Person{
	protected string firstName;
	protected string lastName;
	protected int id;
	
	public Person(){}
	public Person(string firstName, string lastName, int identification){
			this.firstName = firstName;
			this.lastName = lastName;
			this.id = identification;
	}
	public void printPerson(){
		Console.WriteLine("Name: " + lastName + ", " + firstName + "\nID: " + id); 
	}
}

class Student : Person{
    private int[] testScores;  
  
    public Student() {}
    public Student(string firstName, string lastName, int identification, int[] scores){
            this.firstName = firstName;
            this.lastName = lastName;
            this.id = identification;
            this.testScores = scores;
    }

    public char Calculate()
    {
        double average = testScores.Sum(s => s) / testScores.Count();
        if (average >= 90 && average <= 100) return 'O';
        else if (average >= 80 && average < 90) return 'E';
        else if (average >= 70 && average < 80) return 'A';
        else if (average >= 55 && average < 70) return 'P';
        else if (average >= 40 && average < 55) return 'D';
        else return 'T';
    }
}

class Solution {
	static void Main() {
		string[] inputs = Console.ReadLine().Split();
		string firstName = inputs[0];
	  	string lastName = inputs[1];
		int id = Convert.ToInt32(inputs[2]);
		int numScores = Convert.ToInt32(Console.ReadLine());
		inputs = Console.ReadLine().Split();
	  	int[] scores = new int[numScores];
		for(int i = 0; i < numScores; i++){
			scores[i]= Convert.ToInt32(inputs[i]);
		} 
	  	
		Student s = new Student(firstName, lastName, id, scores);
		s.printPerson();
		Console.WriteLine("Grade: " + s.Calculate() + "\n");
	}
}

File: 30 Days of Code/Day 12 Inheritance/Day 12 Inheritance.cpp
#include <iostream>
#include <vector>

using namespace std;


class Person{
	protected:
		string firstName;
		string lastName;
		int id;
	public:
		Person(string firstName, string lastName, int identification){
			this->firstName = firstName;
			this->lastName = lastName;
			this->id = identification;
		}
		void printPerson(){
			cout<< "Name: "<< lastName << ", "<< firstName <<"\nID: "<< id << "\n"; 
		}
	
};

class Student :  public Person{
    private:
        vector<int> testScores;
    public:
        /*    
        *   Class Constructor
        *   
        *   Parameters:
        *   firstName - A string denoting the Person's first name.
        *   lastName - A string denoting the Person's last name.
        *   id - An integer denoting the Person's ID number.
        *   scores - An array of integers denoting the Person's test scores.
        */
        // Write your constructor here
        Student(string firstName, string lastName, int identification, vector<int> scores):Person(firstName, lastName, identification) {
            this->testScores = scores;
        }

        /*    
        *   Function Name: calculate
        *   Return: A character denoting the grade.
        */
        char calculate() {
            int sum = 0, avg = 0;
            char res = 'T';
            for (int i = 0; i < testScores.size(); i++) {
                sum += testScores[i];
            }
            avg = sum / testScores.size();
            if (avg >= 90 && avg <= 100)
                res = 'O';
            else if (avg >= 80 && avg < 90)
                res = 'E';
            else if (avg >= 70 && avg < 80)
                res = 'A';
            else if (avg >= 55 && avg < 70)
                res = 'P';
            else if (avg >= 40 && avg < 55)
                res = 'D';
            return res;
        }
};

int main() {
	string firstName;
  	string lastName;
	int id;
  	int numScores;
	cin >> firstName >> lastName >> id >> numScores;
  	vector<int> scores;
  	for(int i = 0; i < numScores; i++){
	  	int tmpScore;
	  	cin >> tmpScore;
		scores.push_back(tmpScore);
	}
	Student* s = new Student(firstName, lastName, id, scores);
	s->printPerson();
	cout << "Grade: " << s->calculate() << "\n";
	return 0;
}